### PR TITLE
IGNITE-15721 Implemented the ability to use configuration instances as direct proxies to underlying storages.

### PR DESCRIPTION
**Pull request by ibessonov**
https://issues.apache.org/jira/browse/IGNITE-15721

### CODE REVIEW

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/Processor.java [hunk 1]
```diff
@@ -152,6 +152,10 @@ private boolean process0(RoundEnvironment roundEnvironment) {
                     .addModifiers(PUBLIC);
 
             for (VariableElement field : fields) {
+                if (field.getModifiers().contains(STATIC)) {
```
Comment (sashapolo) [inline modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/Processor.java:155]: I think that `fields` method already returns only non-static members

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 2]
```diff
@@ -1012,6 +1011,8 @@ public void dropTable(String name) {
                     }
                 }
             } catch (Throwable t) {
+                assert false;
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:1014]: what is this?

Comment (ibessonov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:1014]: I guess that was the reminder for myself. That exception handling is a complete nonsense.

File: modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/Processor.java (hunk 1 above)
Comment (ibessonov) [reply modules/configuration-annotation-processor/src/main/java/org/apache/ignite/internal/configuration/processor/Processor.java:156]: You're right

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/ConfigurationNode.java [hunk 3]
```diff
@@ -72,20 +79,20 @@
     /**
      * Constructor.
      *
-     * @param prefix     Configuration prefix.
-     * @param key        Configuration key.
-     * @param rootKey    Root key.
-     * @param changer    Configuration changer.
+     * @param keys Configuration keys.
```
Comment (SammyVimes) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/ConfigurationNode.java:82]: I think this should be more elaborate

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicProperty.java [hunk 4]
```diff
@@ -50,28 +61,40 @@
...
+
+        this.injectedNameField = INJECTED_NAME.equals(key);
+        this.internalIdField = INTERNAL_ID.equals(key);
     }
 
     /** {@inheritDoc} */
     @Override
     public T value() {
         if (injectedNameField) {
+            // In this case "refreshValue()" is not of type "T", but an "InnerNode" holding it instead.
```
Comment (SammyVimes) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicProperty.java:91]: I see that it's guarded, but external invariants can be broken (especially in large codebase such as our configuration), probably an assertion would make sense?

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicProperty.java [hunk 5]
```diff
@@ -50,28 +61,40 @@
...
+            // "T" must be a String then, this is guarded by external invariants.
             return (T) ((InnerNode) refreshValue()).getInjectedNameFieldValue();
+        } else if (internalIdField) {
+            // In this case "refreshValue()" is not of type "T", but an "InnerNode" holding it instead.
```
Comment (SammyVimes) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicProperty.java:95]: Same as above

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/DirectProxyAsmGenerator.java [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+    /** Fields from the schema class. */
+    private final List<Field> schemaFields;
+
+    /** Fields from all internal extensions. */
+    private final Collection<Field> internalExtensionsFields;
+
+    /** {@link InternalId} field. */
+    @Nullable
+    private final Field internalIdField;
+    private ClassDefinition classDef;
```
Comment (SammyVimes) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/DirectProxyAsmGenerator.java:102]: Missing newline and javadoc

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/DirectProxyAsmGenerator.java [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+                    ),
+                    methodDef.getThis().getField("changer", DynamicConfigurationChanger.class)
+            ));
+        } else {
+            SchemaClassesInfo fieldSchemaClassInfo = cgen.schemaInfo(schemaField.getType());
+
+            ParameterizedType resultType = typeFromJavaClassName(fieldSchemaClassInfo.directProxyClassName);
+
+            if (isConfigValue(schemaField)) {
+                // new BarDirectProxy(appendKey(this.keys, new KeyPathNode("name")), changer);
```
Comment (SammyVimes) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/DirectProxyAsmGenerator.java:239]: What's Bar?

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java [hunk 8]
```diff
@@ -963,4 +1067,70 @@ public static boolean containsNameAnnotation(Field schemaField) {
...
+     *     var aElementProxy = namedListProxy.get("a");
+     *
+     *     // This operation performs actual reading. It'll throw an exception if element named "a" doesn't exist anymore.
+     *     // It's been renamed or deleted.
+     *     var aElement = aElementProxy.value();
+     *
+     *     // Creates another proxy.
+     *     var aIdElementProxy = getByInternalId(namedListProxy, aId);
+     *
+     *     // This operation performs actual reading as previously stated. But, unlake the access by name, it won't throw an exception in
```
Comment (SammyVimes) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java:1091]:
```suggestion
     *     // This operation performs actual reading as previously stated. But, unlike the access by name, it won't throw an exception in
```

**Review by SammyVimes — CHANGES_REQUESTED** [review]

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/DirectProxyAsmGenerator.java (hunk 6 above)
Comment (tkalkirill) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/DirectProxyAsmGenerator.java:102]: Fix

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicProperty.java (hunk 4 above)
Comment (tkalkirill) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicProperty.java:91]: If it is not used correctly, an **ClassCastException** will be thrown, if it makes sense to add **assert**?

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicProperty.java (hunk 5 above)
Comment (tkalkirill) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicProperty.java:95]: If it is not used correctly, an **ClassCastException** will be thrown, if it makes sense to add **assert**?

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/DirectProxyAsmGenerator.java (hunk 7 above)
Comment (tkalkirill) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/DirectProxyAsmGenerator.java:239]: **ParameterizedType resultType = typeFromJavaClassName(fieldSchemaClassInfo.directProxyClassName);**
Example of direct proxy name

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java (hunk 8 above)
Comment (tkalkirill) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationUtil.java:1091]: fix it

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicProperty.java (hunk 4 above)
Comment (tkalkirill) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicProperty.java:91]: We discussed it in slack - we'll leave it as it is.

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicProperty.java (hunk 5 above)
Comment (tkalkirill) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/DynamicProperty.java:95]: We discussed it in slack - we'll leave it as it is.

**Review by SammyVimes — APPROVED** [review]

**Comment by SammyVimes** [comment]
Merged

**Pull request closed without merging**